### PR TITLE
[Example] Add hardGAT example

### PR DESCRIPTION
**Pull request by Ericcsr**
## Description
Added example of hGAT, hasn't reached state of art performance.

## Potential Issue:
- The some part of computational graph doesn't allow gradient.
- Hyper-parameter Optimization

### CODE REVIEW

**Comment by BarclayII** [comment]
@Ericcsr Which part of the computation does not support gradient?

**Comment by Ericcsr** [comment]
Currently, all part should support gradient flow, and it has been tested

**Comment by Ericcsr** [comment]
> @Ericcsr Which part of the computation does not support gradient?

Currently, all part should support gradient, which has been tested

**Comment by mufeili** [comment]
Can you also add entries in examples/README.md for indexing?

**Comment by sneakerkg** [comment]
@Ericcsr There're some empty files like: benchmarks/benchmarks/__init__.py , please have a clean.

**Comment by sneakerkg** [comment]
Let's change the folder name to hardgat. hgat might be confused with heterogenous gat.

**Review by sneakerkg — CHANGES_REQUESTED** [review]

File: examples/pytorch/hgat/hgat.py [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+            ret = h
+        return ret
+
+class HardGANet(nn.Module):
+    def __init__(self,
+                 g,
+                 num_module,
+                 in_dim,
+                 hid_dim=8,
+                 out_dim=7,
```
[the remark this answers is not in the document]
Comment (Ericcsr) [reply examples/pytorch/hgat/hgat.py]: The 97 feature version is applied for graph classification task, I implement node classification.

File: examples/README.md [hunk 2]
```diff
@@ -19,25 +19,27 @@
...
-| [LINE: Large-scale Information Network Embedding](#line)     |                     | :heavy_check_mark:               |                           |                    | :heavy_check_mark: |
-| [DeepWalk: Online Learning of Social Representations](#deepwalk) |                     | :heavy_check_mark:               |                           |                    | :heavy_check_mark: |
-| [Self-Attention Graph Pooling](#sagpool)                     |                     |                                  | :heavy_check_mark:        |                    |                    |
-| [Convolutional Networks on Graphs for Learning Molecular Fingerprints](#nf) |                     |                                  | :heavy_check_mark:        |                    |                    |
+| [MixHop: Higher-Order Graph Convolutional Architectures via Sparsified Neighborhood Mixing](#mixhop)                     | :heavy_check_mark:  |                                  |                           |                    |                    |
```
[the remark this answers is not in the document]
Comment (Ericcsr) [reply examples/README.md:22]: May be due to an additional space or tab located some where by accident

File: examples/pytorch/hgat/train.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
+"""
+Graph Attention Networks in DGL using SPMV optimization.
```
[the remark this answers is not in the document]
Comment (Ericcsr) [reply examples/pytorch/hgat/train.py:2]: Sure

File: examples/pytorch/hgat/train_ppi.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
+"""
```
[the remark this answers is not in the document]
Comment (Ericcsr) [reply examples/pytorch/hgat/train_ppi.py:1]: OK,sure

File: examples/pytorch/hgat/hgao.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+----------
+Paper: https://arxiv.org/abs/1710.10903
+Author's code: https://github.com/PetarV-/GAT
+Pytorch implementation: https://github.com/Diego999/pyGAT
+"""
+
+import torch
+import torch.nn as nn
+import dgl.function as fn
+from dgl.nn.pytorch import edge_softmax, GATConv
```
[the remark this answers is not in the document]
Comment (Ericcsr) [reply examples/pytorch/hgat/hgao.py:13]: OK, sure

File: examples/pytorch/hgat/hgao.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        # Initialize Dropouts
+        self.feat_drop = nn.Dropout(feat_drop)
+        self.attn_drop = nn.Dropout(attn_drop)
+        self.leaky_relu = nn.LeakyReLU(negative_slope)
+
+        self.reset_parameters()
+        self.activation = activation
+
+    def reset_parameters(self):
+        # TODO: Maybe need exactly same initialization
```
[the remark this answers is not in the document]
Comment (Ericcsr) [reply examples/pytorch/hgat/hgao.py]: Same initialization technique xavier

File: examples/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sneakerkg) [reply examples/README.md:22]: There're conflicts with master now. Please rebase remote master. The issue for those empty files might be resolved as well.

**Review by sneakerkg — CHANGES_REQUESTED** [review]

File: examples/pytorch/hardgat/hgao.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+                                   'causing silent performance regression. '
+                                   'Adding self-loop on the input graph by '
+                                   'calling `g = dgl.add_self_loop(g)` will resolve '
+                                   'the issue. Setting ``allow_zero_in_degree`` '
+                                   'to be `True` when constructing this module will '
+                                   'suppress the check and let the code run.')
+            # projection process to get importance vector y
+            graph.ndata['y'] = torch.abs(torch.matmul(self.p,feat.T).view(-1))/torch.norm(self.p,p=2)
+            # Use edge message passing function to get the weight from src node
+            graph.apply_edges(self.n2e_weight_transfer)
```
Comment (yzh119) [inline examples/pytorch/hardgat/hgao.py:83]: Please use `dgl.function.copy_u` instead.

**Review by sneakerkg — APPROVED** [review]
